packages/pass-style/src/byteArray.js

```diff
@@ -71,3 +72,150 @@ export const ByteArrayHelper = harden({
+ * not be freezable, and would be a read-write view onto `byteArray`.
+ * @returns {Uint8Array} a view into the contents of `byteArray`.
+ */
+export const byteArrayToUint8Array = byteArray => {
+  if (getPrototypeOf(byteArray) === ArrayBuffer.prototype) {
+    // If `byteArray` inherits directly from `ArrayBuffer.prototype`,
+    // then assume `byteArray` is a genuine `ArrayBuffer`,
+    // as it would be in a native implementation
+    // of the Immutable ArrayBuffer proposal. This conversion to a Uint8Array
+    // is then zero-copy. If `byteArray` is indeed an Immutable ArrayBuffer,
+    // then the returned Uint8Array is freezable.
+    return new Uint8Array(byteArray);
+  } else {
+    // Otherwise, assume `byteArray` may be a shim-emulated
+    // Immutable ArrayBuffer. This `.slice()` would then cost a copy into a
+    // fresh mutable normal ArrayBuffer, and the returned Uint8Array would
+    // be a read-write window into that mutable ArrayBuffer.
+    const genuineArrayBuffer = byteArray.slice();
+    return byteArrayToUint8Array(genuineArrayBuffer);
+  }
+};
+harden(byteArrayToUint8Array);
+
+/**
+ * Converts a `Uint8Array` to a `ByteArray`, i.e., a hardened
+ * Immutable ArrayBuffer whose `passStyleOf` is `'byteArray'`.
+ *
+ * @param {Uint8Array} uint8Array
+ * @returns {ArrayBuffer} A ByteArray, i.e., a hardened Immutable ArrayBuffer
+ */
+export const uint8ArrayToByteArray = uint8Array =>
+  // If we're using a native implementation of the full Immutable ArrayBuffer
+  // proposal, and if `uint8Array`'s backing store is such a genuine
+  // Immutable ArrayBuffer, and if the native implementation does the obvious
+  // optimization of `immutableArrayBuffer.sliceToImmutable(0, length)`,
+  // then this conversion is zero-copy.
+  // Otherwise, it pays the copy cost implicit in `sliceToImmutable`.
+  //
+  // @ts-expect-error How can shim add to `ArrayBuffer` ts type?
+  harden(uint8Array.buffer.sliceToImmutable(0, uint8Array.length));
+harden(uint8ArrayToByteArray);
+
+/**
+ * An alternative to `encodeBase64` from the `@endo/base64` package, but for
+ * hex.
+ *
+ * TODO Someday we should deprecate this in favor of some implementation of
+ * https://github.com/tc39/proposal-arraybuffer-base64.
+ * If compat, we should deprecate `@endo/base64` at the same time.
+ *
+ * As with `encodeBase64`, using a JS-based implementation is
+ * ridiculously slow, especially on XS. Unlike `encodeBase64`,
+ * this one does not yet know how to switch to a native implementation
+ * if available.
+ *
+ * Based on `uint8ArrayToHex` from
+ * https://github.com/tc39/proposal-arraybuffer-base64/blob/main/playground/polyfill-core.mjs
+ *
+ * @param {Uint8Array} data
+ * @returns {string} hex encoded string
+ */
+export const uint8ArrayToHex = data => {
+  const len = data.length;
+  const out = [];
+  for (let i = 0; i < len; i += 1) {
+    out.push(data[i].toString(16).padStart(2, '0'));
```

Comment:
It would not be awful to precompute or even embed a 256 entry table.

Reply:
Well, not so done. After the switch, tests that used to succeed are now failing. This might be because the old code was incorrect and the tests followed along, or that the new `hex.js` code is incorrect. I'll track this down.

Reply:
Unresolving